// bfast/dist/esm/bfast.database.js
import { DatabaseController } from "./controllers/database.controller";
import { BulkController } from "./controllers/bulk.controller";
var BfastDatabase = /** @class */ (function () {
    function BfastDatabase(appName, httpClientController, rulesController, authController) {
        this.appName = appName;
        this.httpClientController = httpClientController;
        this.rulesController = rulesController;
        this.authController = authController;
    }
    BfastDatabase.prototype.domain = function (domainName) {
        return new DatabaseController(domainName, this.httpClientController, this.rulesController, this.authController, this.appName);
    };
    BfastDatabase.prototype.collection = function (collectionName) {
        return this.domain(collectionName);
    };
    BfastDatabase.prototype.table = function (tableName) {
        return this.domain(tableName);
    };
    BfastDatabase.prototype.tree = function (name) {
        return this.domain(name);
    };
    // syncs(treeName: string, synced?: (syncs: SyncsController) => void): SyncsController {
    //     return SyncsController.getInstance(
    //         this.appName,
    //         treeName,
    //         getConfig().credential(this.appName).projectId,
    //         this.bulk(),
    //         this.cacheAdapter,
    //         this.tree(treeName),
    //         synced
    //     );
    // }
    BfastDatabase.prototype.bulk = function () {
        return new BulkController(this.appName, this.httpClientController, this.rulesController, this.authController);
    };
    return BfastDatabase;
}());
export { BfastDatabase };
