// bfast/dist/es2015/bfast.database.js
import { DatabaseController } from "./controllers/database.controller";
import { BulkController } from "./controllers/bulk.controller";
export class BfastDatabase {
    constructor(appName, httpClientController, rulesController, authController) {
        this.appName = appName;
        this.httpClientController = httpClientController;
        this.rulesController = rulesController;
        this.authController = authController;
    }
    domain(domainName) {
        return new DatabaseController(domainName, this.httpClientController, this.rulesController, this.authController, this.appName);
    }
    collection(collectionName) {
        return this.domain(collectionName);
    }
    table(tableName) {
        return this.domain(tableName);
    }
    tree(name) {
        return this.domain(name);
    }
    // syncs(treeName: string, synced?: (syncs: SyncsController) => void): SyncsController {
    //     return SyncsController.getInstance(
    //         this.appName,
    //         treeName,
    //         getConfig().credential(this.appName).projectId,
    //         this.bulk(),
    //         this.cacheAdapter,
    //         this.tree(treeName),
    //         synced
    //     );
    // }
    bulk() {
        return new BulkController(this.appName, this.httpClientController, this.rulesController, this.authController);
    }
}
