// bfast/dist/cjs/bfast.database.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BfastDatabase = void 0;
var database_controller_1 = require("./controllers/database.controller");
var bulk_controller_1 = require("./controllers/bulk.controller");
var BfastDatabase = /** @class */ (function () {
    function BfastDatabase(appName, httpClientController, rulesController, authController) {
        this.appName = appName;
        this.httpClientController = httpClientController;
        this.rulesController = rulesController;
        this.authController = authController;
    }
    BfastDatabase.prototype.domain = function (domainName) {
        return new database_controller_1.DatabaseController(domainName, this.httpClientController, this.rulesController, this.authController, this.appName);
    };
    BfastDatabase.prototype.collection = function (collectionName) {
        return this.domain(collectionName);
    };
    BfastDatabase.prototype.table = function (tableName) {
        return this.domain(tableName);
    };
    BfastDatabase.prototype.tree = function (name) {
        return this.domain(name);
    };
    // syncs(treeName: string, synced?: (syncs: SyncsController) => void): SyncsController {
    //     return SyncsController.getInstance(
    //         this.appName,
    //         treeName,
    //         getConfig().credential(this.appName).projectId,
    //         this.bulk(),
    //         this.cacheAdapter,
    //         this.tree(treeName),
    //         synced
    //     );
    // }
    BfastDatabase.prototype.bulk = function () {
        return new bulk_controller_1.BulkController(this.appName, this.httpClientController, this.rulesController, this.authController);
    };
    return BfastDatabase;
}());
exports.BfastDatabase = BfastDatabase;
